### PR TITLE
listing, lookup and getattr optimizations

### PR DESCRIPTION
**Pull request by ploki**
- reuse listing information to speed up readdir+
- avoid getattr after lookup, when possible, reuse attributes from the lookup when performing a getattr
- optimization of getattr, don't perform unconditionally two dbd requests.
- design doc updated

### CODE REVIEW

File: Design.md [hunk 1]
```diff
@@ -297,3 +297,44 @@ The following information is kept up to date by Scality NFS
  - all other fields come from a compile time template
 
 FIXME: now() must be used only when {A,M}TIME_SERVER flags are set.
+
+### Optimized readdir+
+
+Previously, the code did not take benefit from the information returned by
+the listing and operations were decoupled to keep a simple design. This led
+to having:
+ - readdir cost of one roundtrip par 50 entries
```
Comment (LaureVergeron) [inline Design.md]: s/par/per

File: Design.md [hunk 2]
```diff
@@ -297,3 +297,44 @@ The following information is kept up to date by Scality NFS
...
+ - lookup cost of two roundtrips per entry in the listing
+ - getattr cost of one roundtrip per entry int the listing
+
+The lookup had to perform two roundtrip because of the possibility of clashes in names (e.g. an object named "foo" and a directory named "foo/")
```
Comment (LaureVergeron) [inline Design.md]: s/roundtrip/roundtrips

File: Design.md [hunk 3]
```diff
@@ -297,3 +297,44 @@ The following information is kept up to date by Scality NFS
...
+
+During a readdir+, the information gathered by the listing is used to evict
+all lookup.The code now consider the entry as existing and of the right type
```
Comment (LaureVergeron) [inline Design.md]: s/consider/considers

File: Design.md [hunk 4]
```diff
@@ -297,3 +297,44 @@ The following information is kept up to date by Scality NFS
...
+(thanks to commonPrefixes/contents separation). And furthermore, the getattr
+is not accomplished on regular files, thanks to the information from the
+listing which is enough to feed the object attributes.
+On directories, an Metadata request is performed because the commonPrefixes
```
Comment (LaureVergeron) [inline Design.md]: s/an/a/

File: Design.md [hunk 5]
```diff
@@ -297,3 +297,44 @@ The following information is kept up to date by Scality NFS
...
+from the listing may not reflect actual object (remember of placeholders).
```
Comment (LaureVergeron) [inline Design.md]: I don't think you need "of" here.

File: Design.md [hunk 6]
```diff
@@ -297,3 +297,44 @@ The following information is kept up to date by Scality NFS
...
+
+This optimization will only be possible as long as the only object attribute needed for the translation to posix attributes will be mtime and filesize.
```
Comment (LaureVergeron) [inline Design.md]: s/attribute/attributes (mtime+filesize)

File: Design.md [hunk 7]
```diff
@@ -297,3 +297,44 @@ The following information is kept up to date by Scality NFS
...
+
+__Incomplete Attributes__
+
+Files attributes read during optimized listing are incomplete, Part locations
```
Comment (LaureVergeron) [inline Design.md]: Why capitalize part? Also, I think it should be plural

File: Design.md [hunk 8]
```diff
@@ -297,3 +297,44 @@ The following information is kept up to date by Scality NFS
...
+information is missing. Consequently, prior to a data read or a data write,
+a getattr has to be done in order to retrieve the locations of the parts
+constituting the file. 
+
+### Optimized lookup/getattrs
+
+On standalone, lookup/getattr operations, the lookup now performs a first
+exact match lookup to favor performance on files. if it succeed the
```
Comment (LaureVergeron) [inline Design.md]: I'd change favor to enhance for clarity

**Review by LaureVergeron — COMMENTED** [review]
I am not familiar enough with that piece of code that I can comment on the design, but I figured grammar-spotting was better than nothing...

**Comment by ploki** [comment]
Thanks @LaureVergeron

**Pull request merged**